src/DFA/DFA.fs

```diff
@@ -4,96 +4,55 @@ open System.Collections.Generic
 
 open Ast.Common
 
-type State<'a>(id: 'a, isStart: bool, isFinal: bool) =
-    let transitions = Dictionary<bit list, State<'a>>()
+type State = { Name: string; Id: int }
 
-    member this.AddTransition tupleOfChars state =
-        if transitions.ContainsKey(tupleOfChars) then
-            failwithf $"Transition for symbol {tupleOfChars} already exists!"
-
-        transitions.Add(tupleOfChars, state)
-
-    member this.IsStart = isStart
-    member this.IsFinal = isFinal
-    member this.Id = id
-
-    member this.GoesTo char =
-        let exists, v = transitions.TryGetValue char
-        if exists then Some v else None
-
-    override this.ToString() = string this.Id
-
-    member this.GetAllTransitions() =
-        [ for kvp in transitions -> kvp.Key, kvp.Value ]
-
-
-type Result<'a> =
+type Result =
     | Accept
     | Partial of bit list list
-    | Fail of State<'a>
-
-[<Struct>]
-type Configuration<'a> =
-    val CurrentState: State<'a>
-    val RestOfInput: bit list list
+    | Fail of State
 
-    new(state, rest) =
-        { CurrentState = state
-          RestOfInput = rest }
-
-type DFA<'a>(startState: State<'a>) =
-    let rec step (configuration: Configuration<'a>) =
-        match configuration.RestOfInput with
+type DFA(startState: State, finalStates: Set<State>, transitions: Dictionary<State, Dictionary<bit list, State>>) =
```

Comment:
If we don't mutate these dictionaries anywhere, as it seems to me, then we can use `Map`